ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/BlockBroadcaster.java

```diff
@@ -16,21 +16,34 @@
 import tech.pegasys.pantheon.ethereum.eth.manager.EthContext;
 import tech.pegasys.pantheon.ethereum.eth.messages.NewBlockMessage;
 import tech.pegasys.pantheon.ethereum.p2p.api.PeerConnection;
+import tech.pegasys.pantheon.util.Subscribers;
 import tech.pegasys.pantheon.util.uint.UInt256;
 
+import java.util.function.Consumer;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class BlockBroadcaster {
   private static final Logger LOG = LogManager.getLogger();
 
   private final EthContext ethContext;
+  private final Subscribers<Consumer<Block>> newBlockSubscribers = new Subscribers<>();
 
   public BlockBroadcaster(final EthContext ethContext) {
     this.ethContext = ethContext;
   }
 
+  public long subscribePropagateNewBlocks(final Consumer<Block> callback) {
+    return newBlockSubscribers.subscribe(callback);
+  }
+
+  public void unsubscribePropagateNewBlocks(final long id) {
+    newBlockSubscribers.unsubscribe(id);
+  }
+
   public void propagate(final Block block, final UInt256 totalDifficulty) {
+    newBlockSubscribers.forEach(listener -> listener.accept(block));
```

Comment:
I don't think you want to add this functionality here.  `Blockchain.observeBlockAdded` already supports observation of new block events.

Reply:
This event needs to be fired between when we validate the header and when we validate the entire block.

The stack for this code is in part
```
Breakpoint reached
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockBroadcaster.propagate(BlockBroadcaster.java:46)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.broadcastBlock(BlockPropagationManager.java:248)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.lambda$validateAndProcessPendingBlock$6(BlockPropagationManager.java:306)
	  at ...
```

And the stack for using the `Blockchain.observeBlockAdded` event is
```
...
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.runImportTask(BlockPropagationManager.java:323)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.validateAndProcessPendingBlock(BlockPropagationManager.java:307)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager.lambda$importOrSavePendingBlock$5(BlockPropagationManager.java:297)
	  at tech.pegasys.pantheon.ethereum.eth.sync.BlockPropagationManager$$Lambda$1088.740466537.get(Unknown Source:-1)
...
```

Like 307 for BlockPropagationManager runs the full block validation, which is later than the requirements we have for when this information should be pushed to the events plugin.  We don't want full block validation for this event just header (e.g. proof of work) validation.

Reply:
Ok, that makes sense.  I think we just need to change the naming to make it clear what this event is.  In the future if we handle NEW_BLOCK_HASHES / NEW_BLOCK messages differently, we may want to expose more events to get all of the new blocks on the network.  Most clients don't propagate out NEW_BLOCK messages in response to receiving NEW_BLOCK_HASHES from the network, for example.